addons/mass_mailing/models/mailing.py

```diff
@@ -664,12 +664,19 @@ def action_cancel(self):
         self.write({'state': 'draft', 'schedule_date': False, 'schedule_type': 'now', 'next_departure': False})
 
     def action_retry_failed(self):
-        failed_mails = self.env['mail.mail'].sudo().search([
-            ('mailing_id', 'in', self.ids),
-            ('state', '=', 'exception')
-        ])
-        failed_mails.mapped('mailing_trace_ids').unlink()
-        failed_mails.unlink()
+        """ Remove all failed emails and their traces, and try sending them again."""
+        # Use batching to prevent cache overfill in unlink()
+        batch_size = 1000
+        while True:
```

Comment:
Don't like while True + break construct. Instead would have a while(failed_mails), recomputing failed_mails each time, until we have no one left.